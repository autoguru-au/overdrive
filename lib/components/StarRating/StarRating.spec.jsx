import React from 'react';
import { render } from '@testing-library/react';
import { EStarRatingSize, StarRating } from '.';

const getFullStarsNum = el => {
	return el.querySelectorAll('.fullStar').length;
};

const getHalfStarsNum = el => {
	return el.querySelectorAll('.halfStar').length;
};

const getEmptyStarsNum = el => {
	return el.querySelectorAll('.emptyStar').length;
};

describe('<StarRating />', () => {
	it('should not throw', () =>
		expect(() => render(<StarRating />)).not.toThrow());

	it('should match snapshot without value', () => {
		expect(render(<StarRating />).container.firstChild).toMatchSnapshot();
	});

	it('should match snapshot with value', () => {
		expect(
			render(<StarRating rating={2.7} />).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot with label', () => {
		expect(
			render(<StarRating rating={3.8} label="Hello World!" />).container
				.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for small size', () => {
		expect(
			render(
				<StarRating
					rating={3.8}
					label="Hello World!"
					size={EStarRatingSize.Small}
				/>,
			).container.firstChild,
		).toMatchSnapshot();
	});

	it('should pass on className to dom element', () => {
		expect(
			render(<StarRating className="rating-class" />).container
				.firstChild,
		).toHaveClass('rating-class');
	});

	it('should add a span element inside with the rating value if label is not provided value', () => {
		const { container } = render(<StarRating rating={4.1} />);
		expect(container.querySelector('span:nth-child(2)')).toHaveTextContent(
			'4.1',
		);
	});

	it('should add a span element inside with the label text value', () => {
		const { container } = render(<StarRating label="Hello World!" />);
		expect(container.querySelector('span:nth-child(2)')).toHaveTextContent(
			'Hello World!',
		);
	});

	it('should have correct star combination for rating 0', () => {
		const { container: rating } = render(<StarRating rating={0} />);

		expect(getFullStarsNum(rating)).toEqual(0);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(5);
	});

	it('should have correct star combination for rating 1', () => {
		const { container: rating } = render(<StarRating rating={1} />);

		expect(getFullStarsNum(rating)).toEqual(1);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(4);
	});

	it('should have correct star combination for rating 2', () => {
		const { container: rating } = render(<StarRating rating={2} />);

		expect(getFullStarsNum(rating)).toEqual(2);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(3);
	});

	it('should have correct star combination for rating 3', () => {
		const { container: rating } = render(<StarRating rating={3} />);

		expect(getFullStarsNum(rating)).toEqual(3);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(2);
	});

	it('should have correct star combination for rating 4', () => {
		const { container: rating } = render(<StarRating rating={4} />);

		expect(getFullStarsNum(rating)).toEqual(4);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(1);
	});

	it('should have correct star combination for rating 5', () => {
		const { container: rating } = render(<StarRating rating={5} />);

		expect(getFullStarsNum(rating)).toEqual(5);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(0);
	});

	it('should round down to 0 for decimals lower than 0.2', () => {
		const { container: rating } = render(<StarRating rating={3.1} />);

		expect(getFullStarsNum(rating)).toEqual(3);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(2);
	});

	it('should round down to 0 for decimals equal to 0.2', () => {
		const { container: rating } = render(<StarRating rating={4.2} />);

		expect(getFullStarsNum(rating)).toEqual(4);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(1);
	});

	it('should round down to 0.5 for decimals larger than 0.2 and smaller 0.5', () => {
		const { container: rating } = render(<StarRating rating={2.3} />);

		expect(getFullStarsNum(rating)).toEqual(2);
		expect(getHalfStarsNum(rating)).toEqual(1);
		expect(getEmptyStarsNum(rating)).toEqual(2);
	});

	it('should keep 0.5 decimals', () => {
		const { container: rating } = render(<StarRating rating={1.5} />);

		expect(getFullStarsNum(rating)).toEqual(1);
		expect(getHalfStarsNum(rating)).toEqual(1);
		expect(getEmptyStarsNum(rating)).toEqual(3);
	});

	it('should round u to 1 for decimals equal to 0.8', () => {
		const { container: rating } = render(<StarRating rating={4.8} />);

		expect(getFullStarsNum(rating)).toEqual(5);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(0);
	});

	it('should round up to 1 for decimals larger than 0.8', () => {
		const { container: rating } = render(<StarRating rating={3.9} />);

		expect(getFullStarsNum(rating)).toEqual(4);
		expect(getHalfStarsNum(rating)).toEqual(0);
		expect(getEmptyStarsNum(rating)).toEqual(1);
	});
});
