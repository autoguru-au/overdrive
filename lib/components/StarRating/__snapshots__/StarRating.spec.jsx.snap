// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarRating /> should match snapshot for small size 1`] = `
<span
  class="root"
>
  <span
    class="starList"
  >
    <i
      class="root star fullStar"
      style="width: 16px; height: 16px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star fullStar"
      style="width: 16px; height: 16px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star fullStar"
      style="width: 16px; height: 16px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star fullStar"
      style="width: 16px; height: 16px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 16px; height: 16px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
  </span>
  <span
    class="root sizeScale3 label"
  >
    Hello World!
  </span>
</span>
`;

exports[`<StarRating /> should match snapshot with label 1`] = `
<span
  class="root"
>
  <span
    class="starList"
  >
    <i
      class="root star fullStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star fullStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star fullStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star fullStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
  </span>
  <span
    class="root sizeScale4 label"
  >
    Hello World!
  </span>
</span>
`;

exports[`<StarRating /> should match snapshot with value 1`] = `
<span
  class="root"
>
  <span
    class="starList"
  >
    <i
      class="root star fullStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star fullStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star halfStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 15.396V6.095l1.71 4.036 4.38.376-3.323 2.878.995 4.281M22 9.244l-7.19-.617L11.998 2 9.19 8.627 2 9.244l5.454 4.726L5.82 21 12 17.272 18.18 21l-1.635-7.03L22 9.244z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
  </span>
  <span
    class="root sizeScale4 label"
  >
    2.7
  </span>
</span>
`;

exports[`<StarRating /> should match snapshot without value 1`] = `
<span
  class="root"
>
  <span
    class="starList"
  >
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
    <i
      class="root star emptyStar"
      style="width: 20px; height: 20px;"
    >
      <svg
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27l6.18 3.728-1.636-7.03L22 9.244l-7.19-.618L12 1.999 9.19 8.625 2 9.243l5.454 4.726-1.635 7.029L12 17.27z"
        />
      </svg>
    </i>
  </span>
  <span
    class="root sizeScale4 label"
  />
</span>
`;
